packages/dashboard/src/contexts/DashContext/DashProvider.tsx

```diff
@@ -23,46 +36,198 @@ function DashProvider({ children }: DashProviderProps): JSX.Element {
   const initCalled = useRef(false);
+  const stateRef = useRef<State>(state);
+  stateRef.current = state;
 
-  window.devLog({ state });
-  useEffect(() => {
-    async function init() {
-      // This obviates the need for a cleanup callback
-      if (initCalled.current) {
-        return;
+  console.debug({ state });
+
+  const dbHelper = useMemo(
+    () => ({
+      dbPromise: stateRef.current.dbPromise,
+      async has(hash: string) {
+        return !!(await (await this.dbPromise).getKey("Compilation", hash));
+      },
+      async insert(hash: string, compilation: Compilation) {
+        (await this.dbPromise).put("Compilation", {
+          dataHash: hash,
+          data: compilation,
+          timeAdded: Date.now()
+        });
+      },
+      async canInsert() {
+        const { usage, quota } = await navigator.storage.estimate();
+        if (usage && quota) {
+          return (
+            usage / quota < ARBITRARY_DB_MAX_PERCENT &&
+            usage < ARBITRARY_DB_MAX_BYTES
+          );
+        }
+      },
+      async prune() {
+        if (await this.canInsert()) return;
+
+        const dbProxy = await this.dbPromise;
+        const transaction = dbProxy.transaction("Compilation", "readwrite");
+        const index = transaction.store.index("TimeAdded");
+
+        for await (const cursor of index.iterate(null, "next")) {
+          await cursor.delete();
+          if (await this.canInsert()) {
+            break;
+          }
+        }
       }
-      initCalled.current = true;
-
-      window.devLog("Initializing message bus client");
-      // Create message bus client
-      const { host, port, provider } = state;
-      const client = new DashboardMessageBusClient({ host, port });
-      await client.ready();
-      dispatch({ type: "set-client", data: client });
-      window.devLog(`Connected to message bus at ws://${host}:${port}`);
-
-      // Client subscribes to and handles messages
-      const subscription = client.subscribe({});
-      const messageHandler = (lifecycle: ReceivedMessageLifecycle<Message>) =>
-        void dispatch({
-          type: "handle-message",
-          data: { lifecycle, provider }
+    }),
+    []
+  );
+
+  const handleWorkflowCompileResult = useCallback(
+    async (data: WorkflowCompileResult) => {
+      const { compilations } = data;
+
+      if (compilations.length === 0 || !stateRef.current.decoder) return;
+
+      let decoderNeedsUpdate = false;
+      const decoderCompilations = [...stateRef.current.decoderCompilations!];
+      const decoderCompilationHashes = new Set(
+        stateRef.current.decoderCompilationHashes
+      );
+
+      // Iterate over incoming compilations and determine if they are useful to
+      // in-memory decoder or db.
+      // Do not try to optimize by:
+      // - Batch inserting.
+      // - Mirroring some kind of db state.
+      for (const compilation of compilations) {
+        const hash = sha1(compilation);
```

Comment:
How do we know whether we've chosen the correct version of the workflow compile result for the transaction in question? I see that everything is keyed on this sha1 hash of the entire `WorkflowCompileResult`, but that won't allow for matching to a specific version of a given contract, will it?

A better approach might be to store off this hash, but also an index of compliations that's keyed by the hash of the deployed contract bytecode. That would let you verify if you're decoding against the correct/deployed version by hashing the result of `eth_getCode` and checking for it in the database.

This would also allow for less arbitrary cache invalidation strategies, as you could treat the database as an LRU cache and only keep around the most recently used N versions of each contract, where N is ideally some user-configurable value.

Reply:
Yea so the db essentially maps compilation hash to compilation + timestamp.

I see two things here:

> What if workflow-compile changes?

At the moment it's only dependent on the shape of `Compilation`, so it that changes, it's probably time to migrate db.

> The current schema means we can't match, hence decode specific version of a given contract?

So the reason for storing compilations is that one compilation is the unit; you cannot break it into e.g. sources, contracts, etc. and reconstruct it at a later time. That being said, all compilations are fed to one decoder, which means if you write contract A, compile, modify A to be A', compile, the decoder will use compilation for A' to decode.

One caveat as noted by [this comment](https://github.com/trufflesuite/truffle/pull/5646#issuecomment-1294246853) is that in some cases the bytecodes are the same despite having different contracts (like a change in param name), in which case with this week's release the decoder will prefer the "latest" compilation.